gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/FsBenchmark.java

```diff
@@ -0,0 +1,468 @@
+        parseInt(args.getOrDefault("--num-threads", String.valueOf(1))));
+
+    return 0;
+  }
+
+  private void benchmarkRandomRead(
+      FileSystem fs, Path testFile, int numOpen, int readSize, int numReads, int numThreads) {
+    System.out.printf(
+        "Running random read test that reads %d bytes from '%s' file %d times per %d open"
+            + " operations in %d threads%n",
+        readSize, testFile, numReads, numOpen, numThreads);
+
+    List<LongSummaryStatistics> openLatencyNsList = new ArrayList<>(numThreads);
+    List<LongSummaryStatistics> seekLatencyNsList = new ArrayList<>(numThreads);
+    List<LongSummaryStatistics> readLatencyNsList = new ArrayList<>(numThreads);
+    List<LongSummaryStatistics> closeLatencyNsList = new ArrayList<>(numThreads);
+
+    ExecutorService executor = Executors.newFixedThreadPool(numThreads);
+    CountDownLatch initLatch = new CountDownLatch(numThreads);
+    CountDownLatch startLatch = new CountDownLatch(1);
+    CountDownLatch stopLatch = new CountDownLatch(numThreads);
+    List<Future<?>> futures = new ArrayList<>(numThreads);
+    for (int i = 0; i < numThreads; i++) {
+      futures.add(
+          executor.submit(
+              () -> {
+                FileStatus fileStatus = fs.getFileStatus(testFile);
+                long fileSize = fileStatus.getLen();
+                long maxReadPositionExclusive = fileSize - readSize + 1;
+
+                LongSummaryStatistics openLatencyNs = newLongSummaryStatistics(openLatencyNsList);
+                LongSummaryStatistics seekLatencyNs = newLongSummaryStatistics(seekLatencyNsList);
+                LongSummaryStatistics readLatencyNs = newLongSummaryStatistics(readLatencyNsList);
+                LongSummaryStatistics closeLatencyNs = newLongSummaryStatistics(closeLatencyNsList);
+
+                ThreadLocalRandom random = ThreadLocalRandom.current();
+                byte[] readBuffer = new byte[readSize];
+
+                initLatch.countDown();
+                startLatch.await();
+                try {
+                  for (int j = 0; j < numOpen; j++) {
+                    try {
+                      long seekPos = random.nextLong(maxReadPositionExclusive);
+
+                      long openStart = System.nanoTime();
+                      FSDataInputStream input = fs.open(testFile);
+                      openLatencyNs.accept(System.nanoTime() - openStart);
+                      try {
+                        for (int k = 0; k < numReads; k++) {
+                          long seekStart = System.nanoTime();
+                          input.seek(seekPos);
+                          seekLatencyNs.accept(System.nanoTime() - seekStart);
+
+                          long readStart = System.nanoTime();
+                          int numRead = input.read(readBuffer);
+                          readLatencyNs.accept(System.nanoTime() - readStart);
+
+                          if (numRead != readSize) {
```

Comment:
Why this needs an error output? Doesn't this mean that it reached the EOF of something like that?

Reply:
No, it means that because of some reason no data was actually read, which theoretically a bug when you test a random read preformance.